pulsar-common/src/main/java/org/apache/pulsar/common/util/protobuf/ByteBufCodedInputStream.java

```diff
@@ -145,20 +144,22 @@ public void readMessage(final ByteBufMessageBuilder builder, final ExtensionRegi
         buf.writerIndex(writerIdx);
     }
 
+    private static final FastThreadLocal<byte[]> localByteArray = new FastThreadLocal<>();
+
     /** Read a {@code bytes} field value from the stream. */
     public ByteString readBytes() throws IOException {
         final int size = readRawVarint32();
         if (size == 0) {
             return ByteString.EMPTY;
         } else {
-            RecyclableHeapByteBuf heapBuf = RecyclableHeapByteBuf.get();
-            if (size > heapBuf.writableBytes()) {
-                heapBuf.capacity(size);
+            byte[] localBuf = localByteArray.get();
+            if (localBuf == null || localBuf.length < size) {
+                localBuf = new byte[Math.max(size, 1024)];
```

Comment:
so, is there any reason to keep byte[] size at least 1KB?

Reply:
Ideally, we want to minimize the number of times that we have to re-assign the `localBuf`. For example, if we're deserializing strings, we might have different sizes: 16, 18, 25, 29. 

Each time the size increases, we have to re-allocate the temp buffer and GC the old one. 

By starting with 1KB, we should cover most cases and ideally never reallocate it. 

Another strategy could be to start a bit smaller (say 128 or 256 bytes) and when we need to increase, add a 20% padding on top, such that if there are similar sized string soon after, we don't need to reallocate again. What do you think?

Reply:
> Another strategy could be to start a bit smaller (say 128 or 256 bytes) and when we need to increase, add a 20% padding on top

What if very first request is 1KB and rest of them are 256 bytes in that case also we will keep 1KB because we won't reduce it back. right? So, if that is the case then probably below simple thing also would work as we will increase only if request has higher size than require and we will never reduce it.
```
if (localBuf == null || localBuf.length < size) {
    localBuf = new byte[Math.max(size, 1024)];
}
```

But as we have limited number of io threads so, I am fine with either way.

Reply:
correct, but in the above snippet we might have a lot of reallocations before we reach the max-size. 

Since we're already coming down from 16Mb --> 1Kb, I wouldn't worry too much at this point.

Reply:
sure.